Execute Script window statement by statement rather than line by line

lloyddewit opened the pull request:
Replaces PR #8494.

Fixes #7014.
Fixes partially https://github.com/africanmathsinitiative/R-Instat/issues/7989 (allow `=` to be used as an assignment operator).
Fixes partially https://github.com/africanmathsinitiative/R-Instat/issues/8533 (show R code comments in output window, new pipe operator and multi-line statement; plus third block of code).
This change will make 'ctrl-enter' execute the R statement at the cursor (rather than just the line), this should fix the current bugs with multi-line statements.

Ongoing, not yet ready for test/review.

lloyddewit commented:
TODO list (please ignore, note for self):

1. If RScript cannot parse the R code, then give user option to pass R code as string directly to R environment
2. Move caret before executing statement (looks more natural to user)
3. Move caret to next non-blank line, treat white-space lines as blank
4. If statement taking longer than expected to execute, then show warning
5. extra spaces could theoretically mean that wrong statement is executed, so sanity check statement before executing
6. Issue #8500 contains `']`, this cannot currently be parsed by RScript
7. If the statement below (from issue #8509) is at end of script and it has no newline, then comment is appended to output:

![image](https://github.com/africanmathsinitiative/R-Instat/assets/57253949/e4c22fe8-6fcf-4977-b01d-7818cc434de0)

![image](https://github.com/africanmathsinitiative/R-Instat/assets/57253949/870854ee-059b-4631-b67b-3c428aa00f38)

If statement has newline, then output is as expected.

lloyddewit commented:
@rdstern 
This is not yet complete but you are welcome to start testing if you wish.
Thanks

rdstern commented:
@lloyddewit many thanks - will do

rdstern commented:
@lloyddewit I am confused about what runs easily and what is strict?

This first one runs fine with the cursor at the start of each line and pressing run.  Then the second starts - to my mind - identically, but gives me this:

<img width="601" alt="image" src="https://github.com/africanmathsinitiative/R-Instat/assets/11226469/bc84e4fa-2058-4242-b7ae-46b57ff77bc3">

But that is when I try to run the first line, that just gives the library command?

  
  ```
library(agridat)
  data(hanover.whitepine)
  dat <- hanover.whitepine

  libs(lattice)
  # Relatively high male-female interaction in growth comared
  # to additive gene action. Response is more consistent within
  # male progeny than female progeny.
  # with(dat, interaction.plot(female, male, length))
  # with(dat, interaction.plot(male, female, length))
  bwplot(length ~ male|female, data=dat,
         main="hanover.whitepine - length for male:female crosses",
         xlab="Male parent", ylab="Epicotyl length")
  
  # Progeny sums match Becker p 83
  sum(dat$length) # 380.58
  aggregate(length ~  female + male, data=dat, FUN=sum)
  
  # Sum of squares matches Becker p 85
  m1 <- aov(length ~ rep + male + female + male:female, data=dat)
  anova(m1)
  
  # Variance components match Becker p. 85
  libs(lme4)
  libs(lucid)
  m2 <- lmer(length ~ (1|rep) + (1|male) + (1|female) + (1|male:female), data=dat)
  #as.data.frame(lme4::VarCorr(m2))
  vc(m2)
  ##         grp        var1 var2    vcov  sdcor
  ## male:female (Intercept) <NA> 0.1369  0.3699
  ##      female (Intercept) <NA> 0.02094 0.1447
  ##        male (Intercept) <NA> 0.1204  0.3469
  ##         rep (Intercept) <NA> 0.01453 0.1205
  ##    Residual        <NA> <NA> 0.2004  0.4477
  
  # Becker used this value for variability between individuals, within plot
  s2w <- 1.109
  
  # Calculating heritability for individual trees
  s2m <- .120
  s2f <- .0209
  s2mf <- .137
  vp <- s2m + s2f + s2mf + s2w  # variability of phenotypes = 1.3869
  4*s2m / vp # heritability male 0.346
  4*s2f / vp # heritability female 0.06
  2*(s2m+s2f)/vp # heritability male+female .203
  # As shown in the boxplot, heritability is stronger through the
  # males than through the females.

```
And this is the second script:  I must be missing something.  Is there something in the whole script that is being checked and affects how anything is processed in that script?

```
  library(agridat)
  data(harris.wateruse)
  dat <- harris.wateruse

  # Compare to Schabenberger & Pierce, fig 7.23
  libs(latticeExtra)
  useOuterStrips(xyplot(water ~ day|species*age,dat, as.table=TRUE,
                        group=tree, type=c('p','smooth'),
                        main="harris.wateruse 2 species, 2 ages (10 trees each)"))


  # Note that measurements on day 268 are all below the trend line and
  # thus considered outliers.  Delete them.
  dat <- subset(dat, day!=268)


  # Schabenberger figure 7.24
  xyplot(water ~ day|tree,dat, subset=age=="A2" & species=="S2",
         as.table=TRUE, type=c('p','smooth'),
         ylab="Water use profiles of individual trees",
         main="harris.wateruse (Age 2, Species 2)")
  
  # Rescale day for nicer output, and convergence issues, add quadratic term
  dat <- transform(dat, ti=day/100)
  dat <- transform(dat, ti2=ti*ti)

  # Start with a subgroup: age 2, species 2
  d22 <- droplevels(subset(dat, age=="A2" & species=="S2"))

  # ----- Model 1, for subgroup A2,S2

  # First, a fixed quadratic that is common to all trees, plus
  # a random quadratic deviation for each tree.
  
  ## Schabenberger, Output 7.26
  ## proc mixed;
  ##   class tree;
  ##   model water = ti ti*ti / s;
  ##   random intercept ti ti*ti/subject=tree;

  libs(nlme,lucid)
  ## We use pdDiag() to get uncorrelated random effects
  m1n <- nlme::lme(water ~ 1 + ti + ti2, data=d22, na.action=na.omit,
             random = list(tree=nlme::pdDiag(~1+ti+ti2)))
  vc(m1n)
  ##       effect variance    stddev
  ##  (Intercept)   0.2691 0.5188
  ##           ti   0      0.0000144
  ##          ti2   0      0.0000039
  ##     Residual   0.1472 0.3837

  if(0){  
    # Various other models with lme4 & asreml
    
    libs(lme4, lucid)
    m1l <- lmer(water ~ 1 + ti + ti2 + (1|tree) +
                  (0+ti|tree) + (0+ti2|tree), data=d22)
    vc(m1l)
    ##      grp        var1 var2   vcov  sdcor
    ##     tree (Intercept) <NA> 0.2691 0.5188
    ##   tree.1          ti <NA> 0      0
    ##   tree.2         ti2 <NA> 0      0
    ## Residual        <NA> <NA> 0.1472 0.3837


    # Once the overall quadratic trend has been removed, there is not
    # too much evidence for consecutive observations being correlated
    ## d22r <- subset(d22, !is.na(water))
    ## d22r$res <- resid(m1n)
    ## xyplot(res ~ day|tree,d22r,
    ##        as.table=TRUE, type=c('p','smooth'),
    ##        ylab="residual",
    ##        main="harris.wateruse - Residuals of individual trees")
    ## op <- par(mfrow=c(4,3))
    ## tapply(d22r$res, d22r$tree, acf)
    ## par(op)
    
    # ----- Model 2, add correlation of consecutive measurements
    
    ## Schabenberger (page 516) adds correlation.
    ## Note how the fixed quadratic model is on the "ti = day/100" scale
    ## and the correlated observations are on the "day" scale.  The
    ## only impact this has on the fitted model is to increase the
    ## correlation parameter by a factor of 100, which was likely
    ## done to get better convergence.
    
    ## proc mixed data=age2sp2;
    ##   class tree;
    ##   model water = ti ti*ti / s ;
    ##   random intercept /subject=tree s;
    ##   repeated /subject=tree type=sp(exp)(day);

    ## Same as SAS, use ti for quadratic, day for correlation
    m2l <- lme(water ~ 1 + ti + ti2, data=d22,
               random = ~ 1|tree,
               cor = corExp(form=~ day|tree),
               na.action=na.omit)
    m2l # Match output 7.27.  Same fixef, ranef, variances, exp corr

    # vc(m2l)
    ##       effect variance stddev
    ##  (Intercept)   0.2656 0.5154
    ##     Residual   0.1541 0.3926

    # ---





    ## Now use asreml.  When I tried rcov=~tree:exp(ti),
    ## the estimated parameter value was on the 'boundary', i.e. 0.
    ## Changing rcov to the 'day' scale produced a sensible estimate
    ## that matched SAS.
    ## Note: SAS and asreml use different parameterizations for the correlation
    ## SAS uses exp(-d/phi) and asreml uses phi^d.
    ## SAS reports 3.79, asreml reports 0.77, and exp(-1/3.7945) = 0.7683274
    ## Note: normally a quadratic would be included as 'pol(day,2)'

    libs(asreml)
    d22 <- d22[order(d22$tree, d22$day),]
    m2a <- asreml(water ~ 1 + ti + ti2,
                  data=d22,
                  random = ~ tree,
                  residual=~tree:exp(day))
    vc(m2a)
    ##         effect component std.error z.ratio constr
    ##  tree!tree.var    0.2656   0.1301      2      pos
    ##     R!variance    0.1541   0.01611     9.6    pos
    ##      R!day.pow    0.7683   0.04191    18    uncon


    # ----- Model 3. Full model for all species/ages.  Schabenberger p. 518

    ## /* Continuous AR(1) autocorrelations included      */
    ## proc mixed data=wateruse;
    ##   class age species tree;
    ##   model water = age*species age*species*ti age*species*ti*ti / noint s;
    ##   random intercept ti / subject=age*species*tree s;
    ##   repeated / subject=age*species*tree type=sp(exp)(day);
    

    m3l <- lme(water ~ 0 + age:species + age:species:ti + age:species:ti2,
               data=dat, na.action=na.omit,
               random = list(tree=pdDiag(~1+ti)),
               cor = corExp(form=~ day|tree) )

    m3l # Match Schabenberger output 7.27.  Same fixef, ranef, variances, exp corr

    # vc(m3l)
    ##       effect variance stddev
    ##  (Intercept)  0.1549  0.3936
    ##           ti  0.02785 0.1669
    ##     Residual  0.16    0.4

    # --- asreml
    
    dat <- dat[order(dat$tree,dat$day),]
    m3a <- asreml(water ~ 0 + age:species + age:species:ti + age:species:ti2,
                  data=dat,
                  random = ~ age:species:tree + age:species:tree:ti,
                  residual = ~ tree:exp(day) )
    # code for asremlr 4
    ## m3a <- asreml(water ~ 0 + age:species + age:species:ti + age:species:ti2,
    ##               data=dat,
    ##               random = ~ age:species:tree + age:species:tree:ti,
    ##               resid = ~ tree:exp(day) )
    
    # vc(m3a) # Note: day.pow = .8091 = exp(-1/4.7217)
    ##                       effect component std.error z.ratio constr
    ##     age:species:tree!age.var   0.1549   0.07192      2.2    pos
    ##  age:species:tree:ti!age.var   0.02785  0.01343      2.1    pos
    ##                   R!variance   0.16     0.008917    18      pos
    ##                    R!day.pow   0.8091   0.01581     51    uncon

  }
  
```

lloyddewit commented:
> I think I have the problem(s) isolated. There are 2 in this second code:
a) I did need to comment out the if(0){ together with the closing bracket.
b) I also had to comment out lines with a square bracket. for example: dat <- dat[order(dat$tree,dat$day),]

@rdstern yes, you are correct.
Part a is covered by lloyddewit/rscript#19.
Part b is covered by lloyddewit/rscript#18 (note: RScript already parses square brackets but it cannot yet parse `[,` or `,]`).

In order to execute statement-by-statement using ctrl-enter, the whole script needs to be parsed by RScript. If any part of the script can't be parsed, then the user needs to select the part to be run or the user needs to select Run All.

I can see that this could be annoying. To change this, I would need to spend time on improving the back up system. I would prefer to invest the time in RScript so that the backup system is required less and less (hopefully at one point never or extremely rarely).

The question is: could the current approach be merged as-is or would you prefer to wait for a new RScript version before merging?
Could we discuss when we talk on 19 Oct?

rdstern commented:
@lloyddewit I am a believer in merging as soon as possible, particularly as I don't see any obvious regression here in other, seemingly unrelated parts of R-Instat.
Your script windows are looking so good, that I want to include them as soon as you feel prepared for this.
In some ways I am also quite happy that you have these problems!  (Sorry I should rephrase that!)  But this is showing that you are doing something that is complicated and difficult.  And that helps me to understand why the other GUIs have not even tried.

Getting to the practical issues I will be happy for this one to be merged and then just need to add the documentation on detecting scripts with possible problems and then describing the work-arounds in those cases.

I see there will be two types of work around.
a) I can run the whole script, but only in complete chunks.
b) There is code in the script that can't be run, so we need to edit the actual code.

An example of b) used to be when there was `= instead of <-`, used as assignment.  I think that's now ok.  If I understand your message above correctly, then `[, or ,]` might be a current instance of b)?  

I would at least be keen to know what causes b) and minimise or eliminate those.  Perhaps the comma at the start of a multiline is still there, or is that now ok, if I mark the whole section?

I am not even asking for that, before we merge.  

Then I could recommend that users "test" a script initially by trying the library command - as I did.  If that is ok then all is simple.  If not, then they can either run in chunks, or comment out the offending sections.  That's all fine!

lloyddewit commented:
@rdstern Thank you for the positive feedback!

> I see there will be two types of work around.
a) I can run the whole script, but only in complete chunks.
b) There is code in the script that can't be run, so we need to edit the actual code.

> An example of b) used to be when there was = instead of <-, used as assignment. I think that's now ok. If I understand your message above correctly, then [, or ,] might be a current instance of b)?

Some code only works using RScript (e.g. `=` as assignment), and some code only works using the heuristics approach (e.g. key words and `[,` or `,]`). The user normally only needs to edit the actual code if a script contains both types (otherwise we can get the code to run by using the either the heuristics or RScript approach).

There may also be some R statements that do not work using either approach. I think these are mainly related to how R-Instat handles the output. An example is lloyddewit/rscript#17.

As I mentioned before, I think the best approach is to fix these problems in RScript so that the heuristics approach is needed less and less.

> Perhaps the comma at the start of a multiline is still there, or is that now ok, if I mark the whole section?

I think this works using both approaches, do you have a test case that fails?

> `dat <- dat[order(dat$tree, dat$dir), ]`

This only currently works using the heuristics approach. There is no output because it is an assignment statement.

I raised #8568 for commenting blocks of code, and added it to the list in #7989.

I hope this makes sense, let me know if you have any questions.
Thanks!

lloyddewit commented:
@rdstern 
This PR is ready for your testing/approval.
This will be hard to test because it's so open ended. Based on my knowledge of the code and where weaknesses could be, statement-by-statement testing could include:
- Ensuring different types of output is as expected
- Code that can't be parsed by RScript is handled correctly
- Multiple ststements on one line (seprated by `;`)
- Statements split over multiple lines
- Comments at start of statement and/or mixed in with multi-line statements
- Ensuring that cursor moves to start of next statement (handling multiple lines, white space, indents, comments etc)
- Correct statement executes regardless of whether cursor is at start/middle/end of statement; regardless of multiple lines, white space, indents, comments etc.
- empty scripts, single-line scripts (with and without newline at end), large scripts
- statements that take a long time to execute

and whatever else you can think of ... 
Thanks!

P.S. Code that can't be parsed by RScript is covered [here](https://github.com/lloyddewit/RScript/issues). Fixing this will be my next task. If anything is missing, then please let me know

rdstern commented:
@lloyddewit great - will do!

rdstern commented:
Here is my first test code and possible problem?  It is from agricolae ssp.plot (split, split plot analysis, so plot being a plot of land and not a graph!).  It is an example with multiple commands on the same line, that I used before.  It now works.  
However, if I run just the line, by putting to cursor at the start of the line, then it just runs to first command, so up to the semi-colon.  If at the end, then it runs the last command.   If I select the line, then it runs all 3 commands and seemingly correctly.

I'm keen to know whether this is a feature, that I can explain to users, or you consider this a bug you can correct.  I can live with either!

```
# Statistical procedures for agricultural research, pag 143
# Grain Yields of Three Rice Varieties Grown under 
#Three Management practices and Five Nitrogen levels; in a
#split-split-plot design with nitrogen as main-plot, 
#management practice as subplot, and variety as sub-subplot 
#factores, with three replications.
library(agricolae)
f <- system.file("external/ssp.csv", package="agricolae")
ssp<-read.csv(f)
model<-with(ssp,ssp.plot(block,nitrogen,management,variety,yield))
gla<-model$gl.a; glb<-model$gl.b; glc<-model$gl.c
Ea<-model$Ea; Eb<-model$Eb; Ec<-model$Ec
op<-par(mfrow=c(1,3),cex=0.6)
out1<-with(ssp,LSD.test(yield,nitrogen,gla,Ea,console=TRUE))
out2<-with(ssp,LSD.test(yield,management,glb,Eb,console=TRUE))
out3<-with(ssp,LSD.test(yield,variety,glc,Ec,console=TRUE))
plot(out1,xlab="Nitrogen",las=1,variation="IQR")
plot(out2,xlab="Management",variation="IQR")
plot(out3,xlab="Variety",variation="IQR")
# with aov
ssp$block<-factor(ssp$block)
ssp$nitrogen<-factor(ssp$nitrogen)
ssp$management<-factor(ssp$management)
ssp$variety<-factor(ssp$variety)
AOV<-aov(yield ~  block + nitrogen*management*variety + Error(block/nitrogen/management),data=ssp)
summary(AOV)
par(op)
```

Still testing!

rdstern commented:
This second example is a useful script I would like to use in one of the forthcoming presentations.  I am running it in different ways and have some general questions, mainly for information.
I was keen to run statement, by statement.
a) I started by putting the cursor always at the start of a line.  I then noticed it ran the statement, even over multiple lines.  Of course that's what you have been aiming for, all along.  I just didn't appreciate the full significance of it all.
b) I then discovered I can put the cursor anywhere in the statement and it will run the whole statement.
c) I then found that comment lines are defined to essentially be part of the statement that follows.  The's even if I have blank lines between the comments and the subsequent command.

So far so good.

d) I then started at the top and clicked Run.  I found it then went to the next statement.  So I can step through a script just by clicking Run repeatedly.

As above, by only request is to know where I am.  So could we see the cursor.

e) I then started selecting bits.  First I selected a single statement, namely libs(HH).  Here is the result:

<img width="316" alt="image" src="https://github.com/africanmathsinitiative/R-Instat/assets/11226469/45164c22-2594-4e63-a274-cd0dc8ccf7ef">

We don't need that comments repeated twice.  I think we don't need it at all.

f) I then selected the comment above, together with the whole command. The result is here:

<img width="363" alt="image" src="https://github.com/africanmathsinitiative/R-Instat/assets/11226469/c4513275-e0d7-4051-8f56-24c9882f0258">

I am happy that the comment was now added to the output window.  So, again just the 2 extra comment lines to get rid of.

g) Then I selected just the comment.  When I click run, nothing happens.

Could the comment be put into the output window?

h) I then selected part of a statement.  Selecting part of a line often gave an error, as it looked for an object of that name.  Selecting just the command, often gave information about that command.  If I selected a whole first line, it seemed to complete the statement and  gave the result.  I found all those were fine.  So now changes needed.
I also tried selecting the second, or third lines in a multiline statement.  I expected more errors, but it didn't seem to do any harm.  
So can I confirm that selecting the first line of a multiline statement will result in the complete statement being executed.  Selecting other parts of a statement is unpredictable - depends what you select.

i) And the overall good news is that it all seems to run!



 

```
library(agridat)

data(gomez.nitrogen)
dat <- gomez.nitrogen

# Note the depletion of nitrogen over time (stage)
libs(HH)
interaction2wt(nitro ~ rep/trt + trt*stage, data=dat,
               x.between=0, y.between=0,
               main="gomez.nitrogen")
  
# Just the fertilizer profiles
with(dat, interaction.plot(stage, trt, nitro,
                           col=1:4, lty=1:3, main="gomez.nitrogen",
                           xlab="Soil nitrogen at three times for each treatment"))

# Gomez table 6.16
m1 <- aov(nitro ~ Error(rep/trt) + trt*stage, data=dat)
summary(m1)

# Gomez table 6.18
# Treatment                    1 2 3 4 5 6 7 8

cont <- cbind("T7 vs others" = c( 1, 1, 1, 1, 1, 1,-7, 1),
              "T8 vs others"  = c( 1, 1, 1, 1, 1, 1, 0,-6),
              "T2,T5 vs others"   = c(-1, 2,-1,-1, 2,-1, 0, 0),
              "T2 vs T5"       = c( 0, 1, 0, 0,-1, 0, 0, 0))
contrasts(dat$trt) <- cont
contrasts(dat$trt)

m2 <- aov(nitro ~ Error(rep/trt) + trt*stage, data=dat)
summary(m2, expand.split=FALSE,
        split=list(trt=list(
                     "T7 vs others"=1,
                     "T8 vs others"=2,
                     "T2,T5 vs others"=3,
                     "T2 vs T5"=4,
                     rest=c(5,6,7)),
          "trt:stage"=list(
            "(T7 vs others):P"=c(1,8),
            "(T8 vs others):P"=c(2,9),
            "(T2,T5 vs others):P"=c(3,10),
            "(T2 vs T5):P"=c(4,11),
            "rest:P"=c(5,6,7,12,13,14))
          ))


## End(Not run)
```

rdstern commented:
Example 3 - code that used to give a problem:
This is a string, over multiple lines containg hashtags.
It ran previously only if I collected the multiline statement on a single line.
I report that it now runs fine.
Nothing new needed.  


```
library(stringr)
data <- data.frame(
  tweets = c("This is a tweet with #hashtags and #morehashtags.",
  "Another tweet with a single #hashtag.", 
  "No hashtags in this tweet."  ),  
  stringsAsFactors = FALSE)
hashtags = str_extract_all(data$tweets, "#\\S+")
hashtags
```

lloyddewit commented:
@rdstern 
Another area with potential problems is logging. There is:

- the log tab in the Log/Script window
- the automatic save log file
- the output logger which attaches script to output objects (I think @Patowhiz added this)

Please could you test that these are updated as expected by executing script?
Please could you also test that data grids are updated/refreshed when expected?

If the above is difficult to black box test, then should we make it part of the peer review?

Thanks!

rdstern reviewed (dismissed):
Many thanks

Patowhiz commented:
@lloyddewit square brackets are common in R, think of lists.
@rdstern @lloyddewit would you like me to preview this without the brackets feature? Are you happy we merge it without it or should we aim to have that implemented in this PR?

lloyddewit commented:
> square brackets are common in R, think of lists.

@lilyclements 
Thank you for looking at this. 
Square brackets should work in this PR. I think that the only case that fails is when a square bracket is immediately adjacent to an inside comma (`[,` or `,]`). This is covered by lloyddewit/rscript#18.

lloyddewit commented:
> @rdstern @lloyddewit would you like me to preview this without the brackets feature? Are you happy we merge it without it or should we aim to have that implemented in this PR?

@Patowhiz I will fix the square brackets problem in the RScript library, not in this PR. So you are welcome to go ahead and start the peer review.
Thanks

rdstern commented:
@lloyddewit I assume you mean the curly (braces) bracket feature?  I am still happy for it to be merged and for that to come later.  So my Approve still stands.

lloyddewit commented:
@rdstern I fixed the issue with commas inside square bracket operators (e.g. `res <- MCA (poison[,3:8],excl=c(1,3))`). Please could you test?
Thanks

lloyddewit commented:
> @lloyddewit I don't know if it is a right place to report this. When I run this ![image](https://user-images.githubusercontent.com/68591383/275771525-759fc621-d317-443a-8cdb-b53f46926b1e.png) I get the same command in the output ![image](https://user-images.githubusercontent.com/68591383/275771697-f65f7380-c09f-4086-b9de-605a0364072c.png)

@N-thony Thank you for reporting. I think you may be running an older version of this branch. The output from the latest version is shown below. I think this is correct. If I misunderstood, then please let me know.
Thanks

![image](https://github.com/africanmathsinitiative/R-Instat/assets/57253949/a01f22f1-b34e-44d7-9e07-ca76918315ef)

rdstern commented:
@lloyddewit great that seems to work fine now.

I am really enjoying running the script statement by statement, through just pressing the Run repeatedly.

The script below has a number of items of interest!
a) Square brackets - working fine.
b) Simple loop with braces (curly brackets).  I have commented that out in the code.
c) A package we don't have - it is commercial, so I can't simply download it.  Not your problem and it runs everything it can.
d) Comment lines that don't copy to the output.

So I have 2 questions now:
1) Does the editor have a Find/Replace feature?  If so, could it be implemented?  I tried `<ctrl>F` and it loaded an irrelevant dialog!
2) I'm not getting all the comments in the Output window.  I have a very simple solution to this.  I can copy a comment (or anything ) in your script window.  Then I can paste it into (say) a new script window.  But I can't paste it into the bottom of the output window?  Is there a good reason to avoid Patrick adding the facility to paste from the clipboard to the bottom of the output window?



```
  library(agridat)
  data(harris.wateruse)
  dat <- harris.wateruse

  # Compare to Schabenberger & Pierce, fig 7.23
  libs(latticeExtra)
  useOuterStrips(xyplot(water ~ day|species*age,dat, as.table=TRUE,
                        group=tree, type=c('p','smooth'),
                        main="harris.wateruse 2 species, 2 ages (10 trees each)"))


  # Note that measurements on day 268 are all below the trend line and
  # thus considered outliers.  Delete them.
  dat <- subset(dat, day!=268)


  # Schabenberger figure 7.24
  xyplot(water ~ day|tree,dat, subset=age=="A2" & species=="S2",
         as.table=TRUE, type=c('p','smooth'),
         ylab="Water use profiles of individual trees",
         main="harris.wateruse (Age 2, Species 2)")
  
  # Rescale day for nicer output, and convergence issues, add quadratic term
  dat <- transform(dat, ti=day/100)
  dat <- transform(dat, ti2=ti*ti)

  # Start with a subgroup: age 2, species 2
  d22 <- droplevels(subset(dat, age=="A2" & species=="S2"))

  # ----- Model 1, for subgroup A2,S2

  # First, a fixed quadratic that is common to all trees, plus
  # a random quadratic deviation for each tree.
  
  ## Schabenberger, Output 7.26
  ## proc mixed;
  ##   class tree;
  ##   model water = ti ti*ti / s;
  ##   random intercept ti ti*ti/subject=tree;

  libs(nlme,lucid)
  ## We use pdDiag() to get uncorrelated random effects
  m1n <- nlme::lme(water ~ 1 + ti + ti2, data=d22, na.action=na.omit,
             random = list(tree=nlme::pdDiag(~1+ti+ti2)))
  vc(m1n)
  ##       effect variance    stddev
  ##  (Intercept)   0.2691 0.5188
  ##           ti   0      0.0000144
  ##          ti2   0      0.0000039
  ##     Residual   0.1472 0.3837

  #if(0){  
    # Various other models with lme4 & asreml
    
    libs(lme4, lucid)
    m1l <- lmer(water ~ 1 + ti + ti2 + (1|tree) +
                  (0+ti|tree) + (0+ti2|tree), data=d22)
    vc(m1l)
    ##      grp        var1 var2   vcov  sdcor
    ##     tree (Intercept) <NA> 0.2691 0.5188
    ##   tree.1          ti <NA> 0      0
    ##   tree.2         ti2 <NA> 0      0
    ## Residual        <NA> <NA> 0.1472 0.3837


    # Once the overall quadratic trend has been removed, there is not
    # too much evidence for consecutive observations being correlated
    ## d22r <- subset(d22, !is.na(water))
    ## d22r$res <- resid(m1n)
    ## xyplot(res ~ day|tree,d22r,
    ##        as.table=TRUE, type=c('p','smooth'),
    ##        ylab="residual",
    ##        main="harris.wateruse - Residuals of individual trees")
    ## op <- par(mfrow=c(4,3))
    ## tapply(d22r$res, d22r$tree, acf)
    ## par(op)
    
    # ----- Model 2, add correlation of consecutive measurements
    
    ## Schabenberger (page 516) adds correlation.
    ## Note how the fixed quadratic model is on the "ti = day/100" scale
    ## and the correlated observations are on the "day" scale.  The
    ## only impact this has on the fitted model is to increase the
    ## correlation parameter by a factor of 100, which was likely
    ## done to get better convergence.
    
    ## proc mixed data=age2sp2;
    ##   class tree;
    ##   model water = ti ti*ti / s ;
    ##   random intercept /subject=tree s;
    ##   repeated /subject=tree type=sp(exp)(day);

    ## Same as SAS, use ti for quadratic, day for correlation
    m2l <- lme(water ~ 1 + ti + ti2, data=d22,
               random = ~ 1|tree,
               cor = corExp(form=~ day|tree),
               na.action=na.omit)
    m2l # Match output 7.27.  Same fixef, ranef, variances, exp corr

    # vc(m2l)
    ##       effect variance stddev
    ##  (Intercept)   0.2656 0.5154
    ##     Residual   0.1541 0.3926

    # ---

    ## Now use asreml.  When I tried rcov=~tree:exp(ti),
    ## the estimated parameter value was on the 'boundary', i.e. 0.
    ## Changing rcov to the 'day' scale produced a sensible estimate
    ## that matched SAS.
    ## Note: SAS and asreml use different parameterizations for the correlation
    ## SAS uses exp(-d/phi) and asreml uses phi^d.
    ## SAS reports 3.79, asreml reports 0.77, and exp(-1/3.7945) = 0.7683274
    ## Note: normally a quadratic would be included as 'pol(day,2)'

    libs(asreml)
    d22 <- d22[order(d22$tree, d22$day),]
    m2a <- asreml(water ~ 1 + ti + ti2,
                  data=d22,
                  random = ~ tree,
                  residual=~tree:exp(day))
    vc(m2a)
    ##         effect component std.error z.ratio constr
    ##  tree!tree.var    0.2656   0.1301      2      pos
    ##     R!variance    0.1541   0.01611     9.6    pos
    ##      R!day.pow    0.7683   0.04191    18    uncon


    # ----- Model 3. Full model for all species/ages.  Schabenberger p. 518

    ## /* Continuous AR(1) autocorrelations included      */
    ## proc mixed data=wateruse;
    ##   class age species tree;
    ##   model water = age*species age*species*ti age*species*ti*ti / noint s;
    ##   random intercept ti / subject=age*species*tree s;
    ##   repeated / subject=age*species*tree type=sp(exp)(day);
    

    m3l <- lme(water ~ 0 + age:species + age:species:ti + age:species:ti2,
               data=dat, na.action=na.omit,
               random = list(tree=pdDiag(~1+ti)),
               cor = corExp(form=~ day|tree) )

    m3l # Match Schabenberger output 7.27.  Same fixef, ranef, variances, exp corr

    # vc(m3l)
    ##       effect variance stddev
    ##  (Intercept)  0.1549  0.3936
    ##           ti  0.02785 0.1669
    ##     Residual  0.16    0.4

    # --- asreml
    
    dat <- dat[order(dat$tree,dat$day),]
    m3a <- asreml(water ~ 0 + age:species + age:species:ti + age:species:ti2,
                  data=dat,
                  random = ~ age:species:tree + age:species:tree:ti,
                  residual = ~ tree:exp(day) )
    # code for asremlr 4
    ## m3a <- asreml(water ~ 0 + age:species + age:species:ti + age:species:ti2,
    ##               data=dat,
    ##               random = ~ age:species:tree + age:species:tree:ti,
    ##               resid = ~ tree:exp(day) )
    
    # vc(m3a) # Note: day.pow = .8091 = exp(-1/4.7217)
    ##                       effect component std.error z.ratio constr
    ##     age:species:tree!age.var   0.1549   0.07192      2.2    pos
    ##  age:species:tree:ti!age.var   0.02785  0.01343      2.1    pos
    ##                   R!variance   0.16     0.008917    18      pos
    ##                    R!day.pow   0.8091   0.01581     51    uncon

  #}
  
```

rdstern reviewed (approved):
@Patowhiz hope you can check this.  It is still "work in progress" but good for you to examine, and then consider the remaining work as a new later pull request.

lloyddewit commented:
@rdstern Thank you for testing.

> 1. Does the editor have a Find/Replace feature?

No, this is already requested in issue #7989.

> 2. I'm not getting all the comments in the Output window.

I tested and, as far as I can see, all the comments are displayed in the output window. The only exception is when a comment is not followed by an executable statement (i.e. comments at the end of the script); or when only a comment (with no executable R script) is selected and run. These comments are ignored (deliberate design choice). Would you like comments without statements to also be sent to the Output window?
As a temporary workaround you can just add a minimal dummy statement (e.g. a `1`) after the final comment, and execute that (see below).

If I misunderstood anything, then please let me know.
Thanks

![image](https://github.com/africanmathsinitiative/R-Instat/assets/57253949/55f297d9-013b-4dc9-a950-fdf2c4c1cb1c)

![image](https://github.com/africanmathsinitiative/R-Instat/assets/57253949/ae9f5859-2042-4fd5-89df-e001848b81ee)

rdstern commented:
@lloyddewit you seem to imply that would be easy to do.  When I mark and run a comment then yes, it would be great if it could be sent to the output window.

Meantime I am happy with your dummy command.
The other comment that is not sent to the output window is when it is on the same line as a command that produces a result.  I'll give an example below.  Again, I am quite ok with that, but if I deliberately copy and run that comment, then (again) it would be nice if that simply copied to the output window.

Just a minute, while I find it!

lloyddewit commented:
@rdstern 
I could not recreate the commenting problem (see below). Please could you demonstrate it in our meeting?
Thanks

![image](https://github.com/africanmathsinitiative/R-Instat/assets/57253949/91197ca8-8509-4be3-91d7-c8f5c61eecc1)

rdstern commented:
@lloyddewit after our discussion it is clear I made a mistake on that point, i.e. on the comment not appearing sometimes when it was at the end of an executable line.  I had expected it to appear after the result.  Instead it does appear, but above where I looked, as part of the command line, and that is quite right!

lloyddewit commented:
@rdstern You reported that comments at the end of scripts, that are not associated with any R statement, are not shown in the output window. On 17 Oct, you gave this example:

![image](https://github.com/africanmathsinitiative/R-Instat/assets/57253949/afb4ef37-eb98-405a-904d-455ebc0d9bb1)

I have now changed this to what you requested. You can use ctrl-enter to execute comments appended to a script; or you can highlight/execute just a comment. Run All will also show any final script comments in the output window.

I also fixed the bug shown below.

![image](https://github.com/africanmathsinitiative/R-Instat/assets/57253949/283233b5-846a-46c0-8fc9-92f0b5198b81)

Please could you test/approve?
Thanks

lloyddewit commented:
@rdstern I also marked the following issue #7989 task as done. Please could you test that you agree? Thanks

![image](https://github.com/africanmathsinitiative/R-Instat/assets/57253949/dc6ed6a6-5849-4496-bb41-f06d1a943d5d)

rdstern reviewed (approved):
@Patowhiz this still all seems good.
Note the issue of the log window opening is fixed in Antoine's pull request.

Patowhiz reviewed (changes requested):
@lloyddewit thanks for this.
Below are my few requests and suggestions.

lloyddewit commented on instat/clsRLink.vb:
```diff
@@ -715,6 +715,162 @@ Public Class RLink
+            Else
+                Evaluate(strRStatement, bSilent:=False, bSeparateThread:=False,
+                         bShowWaitDialogOverride:=Nothing)
+            End If
```
@Patowhiz Valuable comment - thank you. I tried to fix in the commit below. Please review.

Patowhiz replied:
Looks good.

lloyddewit commented on instat/clsRLink.vb:
```diff
@@ -715,6 +715,162 @@ Public Class RLink
+        Next
+
+        frmMain.UpdateAllGrids()
+    End Sub
```
@Patowhiz yes, you are correct. If you are happy with the commit below, then I will refactor the 2 functions to eliminate duplicate code.
Thanks

Patowhiz replied:
Looks good.

Patowhiz reviewed (commented):
@lloyddewit thanks for addressing my suggestions.
Once you refactor the functions to avoid the duplication we can merge.

lloyddewit commented:
> @lloyddewit thanks for addressing my suggestions. Once you refactor the functions to avoid the duplication we can merge.

@Patowhiz Thanks. I replaced some of the duplicated code with shared private functions. I tried to refactor out more of the duplicated code but found that to make it work, I would need to significantly refactor the existing `RunScript()` function. This would be a significant task. As you know, this function is called by all the dialogs so there would be a high risk of regression. Maybe we should do this one day but I prefer not to add such a significant task to this PR. 
The remaining duplication is mainly between `RunScript()` and `RunScriptFromWindow()`. The latter function is only a backup and is only called when RScript raises an exception. I am working on solving these exceptions in the RScript library. We should be able to remove `RunScriptWindow()` in the next weeks/months. So if we leave the refactoring at the current level, then I don't think we are creating technical debt.
What's your view?

rdstern reviewed (approved):
I'm approving again.  @Patowhiz hope it is ok to merge now, so we can proceed with #8597

Patowhiz commented:
@rdstern @lloyddewit give me more time to look into this. 
I'm still looking into PR #8597 and it's issues, which are not as trivial as @rdstern would have wished them to be.

ChrisMarsh82 commented:
@Patowhiz Can this be merged as the code looks fine? Then if there are more changes they can be done on #8597
